File: server/app/schemas/clip.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

ClipExportStatus = Literal["none", "exporting", "completed", "failed"]


class ClipCreate(BaseModel):
    start_time_sec: float
    end_time_sec: float
    label: str = ""
    notes: str = ""
    anchor_marker_id: str | None = None


class ClipUpdate(BaseModel):
    start_time_sec: float | None = None
    end_time_sec: float | None = None
    label: str | None = None
    notes: str | None = None
    anchor_marker_id: str | None = None
    export_status: ClipExportStatus | None = None
    exported_file_path: str | None = None


class ClipRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    start_time_sec: float
    end_time_sec: float
    label: str
    notes: str
    anchor_marker_id: str | None
    export_status: str
    exported_file_path: str | None
    created_at: datetime

File: BadmFrame/BadmFrame/Models/Clip.swift
import Foundation
import SwiftData

@Model
final class Clip {
    var startTimeSec: Double = 0
    var endTimeSec: Double = 0
    var label: String = ""
    var notes: String = ""
    var anchorMarkerId: String?
    var exportStatusRaw: String = ClipExportStatus.none.rawValue
    var exportedFilePath: String?
    var createdAt: Date = Date()

    var project: Project?

    enum ClipExportStatus: String {
        case none = "none"
        case exporting = "exporting"
        case completed = "completed"
        case failed = "failed"
    }

    init(startTimeSec: Double, endTimeSec: Double, label: String = "", notes: String = "", anchorMarkerId: String? = nil) {
        self.startTimeSec = startTimeSec
        self.endTimeSec = endTimeSec
        self.label = label
        self.notes = notes
        self.anchorMarkerId = anchorMarkerId
        self.createdAt = Date()
    }

    var durationSec: Double { endTimeSec - startTimeSec }

    var startTimeDisplay: String { startTimeSec.preciseDisplayString }
    var endTimeDisplay: String { endTimeSec.preciseDisplayString }
    var durationDisplay: String { durationSec.preciseDisplayString }

    var exportStatus: ClipExportStatus {
        get { ClipExportStatus(rawValue: exportStatusRaw) ?? .none }
        set { exportStatusRaw = newValue.rawValue }
    }
}
